File: backend/app/models/shopping_list.py
"""Shopping list models for the shopping list generation feature."""

from datetime import date as date_type
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field, field_validator


class IngredientCategory(str, Enum):
    """Enumeration of ingredient categories for shopping list grouping."""

    LEGUMES = "legumes"
    VIANDES = "viandes"
    POISSONS = "poissons"
    PRODUITS_LAITIERS = "produits_laitiers"
    EPICERIE = "epicerie"
    BOISSONS = "boissons"
    SURGELES = "surgeles"
    AUTRES = "autres"


class ShoppingListItemBase(BaseModel):
    """Base model for shopping list item fields."""

    ingredient_name: str = Field(
        ..., min_length=1, max_length=255, description="Name of the ingredient"
    )
    quantity: str = Field(
        ..., min_length=1, max_length=100, description="Quantity with unit"
    )
    category: IngredientCategory = Field(
        ..., description="Category for grouping in shopping list"
    )


class ShoppingListItemCreate(ShoppingListItemBase):
    """Request model for creating a shopping list item."""

    week_start: date_type = Field(
        ..., description="Monday of the week (YYYY-MM-DD)"
    )

    @field_validator("week_start")
    @classmethod
    def week_start_must_be_monday(cls, v: date_type) -> date_type:
        """Validate that week_start is a Monday."""
        if v.weekday() != 0:
            raise ValueError("week_start must be a Monday")
        return v


class ShoppingListItemResponse(BaseModel):
    """Response model for a single shopping list item."""

    id: str
    user_id: str
    ingredient_name: str
    quantity: str
    category: IngredientCategory
    is_checked: bool
    week_start: date_type
    created_at: datetime
    updated_at: datetime


class ShoppingListGenerateResponse(BaseModel):
    """Response model for shopping list generation."""

    items: list[ShoppingListItemResponse]
    week_start: date_type
    generated_at: datetime


class CategoryGroupResponse(BaseModel):
    """Response model for items grouped by category."""

    category: IngredientCategory
    items: list[ShoppingListItemResponse]
    item_count: int


class ShoppingListResponse(BaseModel):
    """Response model for the full shopping list grouped by category."""

    categories: list[CategoryGroupResponse]
    week_start: date_type
    total_items: int
    checked_count: int


class ToggleItemResponse(BaseModel):
    """Response model for toggling an item's checked state."""

    item: ShoppingListItemResponse
    is_checked: bool


class UncheckAllResponse(BaseModel):
    """Response model for unchecking all items."""

    count: int
    message: str

File: mobile/types/shopping-list.ts
/**
 * Shopping List Types
 *
 * TypeScript types for the shopping list generation feature.
 */

/**
 * Ingredient category for grouping
 */
export type IngredientCategory =
  | 'legumes'
  | 'viandes'
  | 'poissons'
  | 'produits_laitiers'
  | 'epicerie'
  | 'boissons'
  | 'surgeles'
  | 'autres';

/**
 * Display names for categories (French)
 */
export const CATEGORY_LABELS: Record<IngredientCategory, string> = {
  legumes: 'Legumes',
  viandes: 'Viandes',
  poissons: 'Poissons',
  produits_laitiers: 'Produits laitiers',
  epicerie: 'Epicerie',
  boissons: 'Boissons',
  surgeles: 'Surgeles',
  autres: 'Autres',
};

/**
 * Order of categories for display
 */
export const CATEGORY_ORDER: IngredientCategory[] = [
  'legumes',
  'viandes',
  'poissons',
  'produits_laitiers',
  'epicerie',
  'boissons',
  'surgeles',
  'autres',
];

/**
 * Shopping list item from API
 */
export interface ShoppingListItem {
  id: string;
  user_id: string;
  ingredient_name: string;
  quantity: string;
  category: IngredientCategory;
  is_checked: boolean;
  week_start: string;
  created_at: string;
  updated_at: string;
}

/**
 * Category group with items
 */
export interface CategoryGroup {
  category: IngredientCategory;
  items: ShoppingListItem[];
  item_count: number;
}

/**
 * Response for GET /api/shopping-list
 */
export interface ShoppingListResponse {
  categories: CategoryGroup[];
  week_start: string;
  total_items: number;
  checked_count: number;
}

/**
 * Response for POST /api/shopping-list/generate
 */
export interface ShoppingListGenerateResponse {
  items: ShoppingListItem[];
  week_start: string;
  generated_at: string;
}

/**
 * Response for PUT /api/shopping-list/uncheck-all
 */
export interface UncheckAllResponse {
  count: number;
  message: string;
}
